standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java

```diff
@@ -2247,31 +2247,64 @@ public ValidTxnWriteIdList getValidWriteIds(Long currentTxnId, List<String> tabl
 
   @Override
   public long openTxn(String user) throws TException {
-    OpenTxnsResponse txns = openTxns(user, 1);
+    OpenTxnsResponse txns = openTxnsIntr(user, null, null, 1);
     return txns.getTxn_ids().get(0);
   }
 
+  @Override
+  public List<Long> replOpenTxn(String replPolicy, Iterator<Long> srcTxnIds, int numTxns) throws TException {
+    // As this is called from replication task, the user field is not used.
+    OpenTxnsResponse txns = openTxnsIntr(null, replPolicy, srcTxnIds, numTxns);
+    return txns.getTxn_ids();
+  }
+
   @Override
   public OpenTxnsResponse openTxns(String user, int numTxns) throws TException {
+    return openTxnsIntr(user, null, null, numTxns);
+  }
+
+  private OpenTxnsResponse openTxnsIntr(String user, String replPolicy, Iterator<Long> srcTxnIds,
+                                        int numTxns) throws TException {
     String hostname = null;
     try {
       hostname = InetAddress.getLocalHost().getHostName();
     } catch (UnknownHostException e) {
       LOG.error("Unable to resolve my host name " + e.getMessage());
       throw new RuntimeException(e);
     }
-    return client.open_txns(new OpenTxnRequest(numTxns, user, hostname));
+    OpenTxnRequest rqst = new OpenTxnRequest(numTxns, user, hostname);
+    if (replPolicy != null) {
+      // need to set this only for replication tasks
+      rqst.setReplPolicy(replPolicy);
+      rqst.setReplSrcTxnId(Lists.newArrayList(srcTxnIds));
+    }
+    return client.open_txns(rqst);
+  }
+
+  @Override
+  public void rollbackTxn(long txnid, String replPolicy) throws NoSuchTxnException, TException {
```

Comment:
Here txnId refers to srcTxnId if replPolicy is not-null. If null, then it means target txn id.
I suggest to have separate arguments for both to be more clear. Also, assert inputs which is expected to null etc.
Same comment for commitTxn as well.

Reply:
i am not sure how does it matter if its source or target txn id ..for rollback and commit ..i feel we can keep only one parameter

Reply:
anyways there are two apis one with replpolicy and one without replPolicy ..so changed the name to srcTxnid where replPolicy is present